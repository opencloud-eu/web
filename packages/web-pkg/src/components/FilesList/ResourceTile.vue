<template>
  <div
    ref="observerTarget"
    class="oc-tile-card oc-card oc-card-default"
    :data-item-id="resource.id"
    :class="{
      'oc-tile-card-selected': isResourceSelected,
      'oc-tile-card-disabled': isResourceDisabled && !isProjectSpaceResource(resource),
      'state-trashed': isResourceDisabled && isProjectSpaceResource(resource)
    }"
    @contextmenu="$emit('contextmenu', $event)"
  >
    <div v-if="isHidden" class="oc-tile-card-lazy-shimmer"></div>
    <template v-else>
      <resource-link
        class="oc-card-media-top oc-flex oc-flex-center oc-flex-middle oc-m-rm"
        :resource="resource"
        :link="resourceRoute"
        :is-resource-clickable="isResourceClickable"
        tabindex="-1"
        @click="$emit('click')"
      >
        <div class="oc-tile-card-selection">
          <div v-if="isLoading" class="oc-tile-card-loading-spinner oc-m-s">
            <oc-spinner :aria-label="$gettext('File is being processed')" />
          </div>
          <slot v-else name="selection" :item="resource" />
        </div>
        <oc-tag
          v-if="isResourceDisabled && isProjectSpaceResource(resource)"
          class="resource-disabled-indicator oc-position-absolute"
          type="span"
        >
          <span v-text="$gettext('Disabled')" />
        </oc-tag>
        <div
          v-oc-tooltip="tooltipLabelIcon"
          class="oc-tile-card-preview oc-flex oc-flex-middle oc-flex-center"
          :aria-label="tooltipLabelIcon"
        >
          <slot name="imageField" :item="resource">
            <oc-image v-if="resource.thumbnail" class="tile-preview" :src="resource.thumbnail" />
            <resource-icon
              v-else
              :resource="resource"
              :size="resourceIconSize"
              class="tile-default-image oc-pt-xs"
            >
              <template v-if="showStatusIcon" #status>
                <oc-icon v-bind="statusIconAttrs" size="xsmall" />
              </template>
            </resource-icon>
          </slot>
        </div>
      </resource-link>
      <div class="oc-card-body oc-p-s">
        <div class="oc-flex oc-flex-between oc-flex-middle">
          <div class="oc-flex oc-flex-middle oc-text-truncate resource-name-wrapper">
            <resource-list-item
              :resource="resource"
              :is-icon-displayed="false"
              :is-extension-displayed="isExtensionDisplayed"
              :is-resource-clickable="isResourceClickable"
              :link="resourceRoute"
              @click="$emit('click')"
            />
          </div>
          <div class="oc-flex oc-flex-middle">
            <!-- Slot for indicators !-->
            <slot name="indicators" :item="resource" class="resource-indicators" />
            <!-- Slot for individual actions -->
            <slot name="actions" :item="resource" />
            <!-- Slot for contextmenu -->
            <slot name="contextMenu" :item="resource" />
          </div>
        </div>
        <p v-if="resourceDescription" class="oc-text-left oc-my-rm oc-text-truncate">
          <small v-text="resourceDescription" />
        </p>
      </div>
    </template>
  </div>
</template>

<script setup lang="ts">
import { computed, customRef, ref, unref } from 'vue'
import ResourceIcon from './ResourceIcon.vue'
import ResourceListItem from './ResourceListItem.vue'
import ResourceLink from './ResourceLink.vue'
import { isProjectSpaceResource, Resource } from '@opencloud-eu/web-client'
import { useGettext } from 'vue3-gettext'
import { isSpaceResource } from '@opencloud-eu/web-client'
import { RouteLocationRaw } from 'vue-router'
import { useIsVisible } from '@opencloud-eu/design-system/composables'
import { SizeType } from '@opencloud-eu/design-system/helpers'

const {
  resource,
  resourceRoute,
  isResourceSelected = false,
  isResourceClickable = true,
  isResourceDisabled = false,
  isExtensionDisplayed = true,
  resourceIconSize = 'xlarge',
  lazy = false,
  isLoading = false
} = defineProps<{
  resource?: Resource
  resourceRoute?: RouteLocationRaw
  isResourceSelected?: boolean
  isResourceClickable?: boolean
  isResourceDisabled?: boolean
  isExtensionDisplayed?: boolean
  resourceIconSize?: SizeType
  lazy?: boolean
  isLoading?: boolean
}>()

const emit = defineEmits<{
  (e: 'click'): void
  (e: 'contextmenu', event: MouseEvent | KeyboardEvent): void
  (e: 'itemVisible'): void
}>()

defineSlots<{
  actions?: (props: { item: Resource }) => unknown
  contextMenu?: (props: { item: Resource }) => unknown
  imageField?: (props: { item: Resource }) => unknown
  indicators?: (props: { item: Resource }) => unknown
  selection?: (props: { item: Resource }) => unknown
}>()

const { $gettext } = useGettext()

const observerTarget = customRef((track, trigger) => {
  let $el: HTMLElement
  return {
    get() {
      track()
      return $el
    },
    set(value) {
      $el = value
      trigger()
    }
  }
})

const showStatusIcon = computed(() => {
  return resource.locked || resource.processing
})

const statusIconAttrs = computed(() => {
  if (resource.locked) {
    return {
      name: 'lock',
      fillType: 'fill' as const
    }
  }

  if (resource.processing) {
    return {
      name: 'loop-right',
      fillType: 'line' as const
    }
  }

  return {}
})

const tooltipLabelIcon = computed(() => {
  if (resource.locked) {
    return $gettext('This item is locked')
  }
  return null
})
const resourceDescription = computed(() => {
  if (isSpaceResource(resource)) {
    return resource.description
  }
  return ''
})

const { isVisible } = lazy
  ? useIsVisible({
      target: observerTarget,
      onVisibleCallback: () => emit('itemVisible')
    })
  : { isVisible: ref(true) }

const isHidden = computed(() => !unref(isVisible))

if (!lazy) {
  emit('itemVisible')
}
</script>

<style lang="scss">
.oc-tile-card {
  background-color: var(--oc-role-surface-container) !important;
  box-shadow: none;
  height: 100%;
  display: flex;
  flex-flow: column;
  outline: 0.5px solid var(--oc-role-outline-variant);

  &-disabled {
    pointer-events: none;
    opacity: 0.7;
    filter: grayscale(0.6);

    // Show tooltip on status indicators without handler
    span.oc-status-indicators-indicator {
      pointer-events: all;
    }
  }

  &-loading-spinner {
    z-index: 99;
  }

  &.state-trashed {
    .tile-image,
    .tile-default-image > svg {
      filter: grayscale(100%);
      opacity: 80%;
    }
  }

  .tile-default-image {
    position: relative;
  }

  .oc-card-media-top {
    position: relative;
    aspect-ratio: 16/9;
    justify-content: center;
    width: 100%;

    .oc-tag {
      color: var(--oc-role-on-surface);

      &.resource-disabled-indicator {
        z-index: 1;
      }
    }

    .tile-preview {
      aspect-ratio: 16/9;
      height: 100%;
      object-fit: cover;
      width: 100%;
      border-top-left-radius: 5px;
      border-top-right-radius: 5px;
    }
  }

  &-selected,
  &:hover {
    .oc-tile-card-preview {
      width: calc(100% - var(--oc-space-medium));
      height: calc(100% - var(--oc-space-medium));

      .tile-preview {
        border-radius: 5px !important;
      }
    }
  }
  &:hover {
    background-color: var(--oc-role-secondary-container) !important;
  }
  &-selected {
    background-color: var(--oc-role-surface-container-high) !important;
    outline: 2px solid var(--oc-role-outline);
  }

  &-selection {
    z-index: 1;
    position: absolute;
    top: 0;
    left: 0;

    input {
      background-color: var(--oc-role-surface-container);
    }
  }

  &-preview {
    position: absolute;
    height: 100%;
    width: 100%;
    text-align: center;
  }

  .resource-name-wrapper {
    color: var(--oc-role-on-surface);
    max-width: 70%;
    overflow: hidden;
  }

  &-lazy-shimmer {
    height: 120px;
    opacity: 0.2;
    position: relative;
    overflow: hidden;
  }

  &-lazy-shimmer::after {
    animation: shimmer 2s infinite;
    background-image: linear-gradient(
      90deg,
      rgba(#4c5f79, 0) 0,
      rgba(#4c5f79, 0.2) 20%,
      rgba(#4c5f79, 0.5) 60%,
      rgba(#4c5f79, 0)
    );
    bottom: 0;
    content: '';
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
  }

  @keyframes shimmer {
    100% {
      transform: translateX(100%);
    }
  }
}
</style>
