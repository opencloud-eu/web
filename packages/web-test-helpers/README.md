# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
