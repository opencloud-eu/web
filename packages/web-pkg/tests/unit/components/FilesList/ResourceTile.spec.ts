import { defaultPlugins, shallowMount } from '@opencloud-eu/web-test-helpers'
import ResourceTile from '../../../../src/components/FilesList/ResourceTile.vue'

const getSpaceMock = (disabled = false) => ({
  name: 'Space 1',
  path: '',
  type: 'space',
  isFolder: true,
  disabled,
  getDriveAliasAndItem: () => '1'
})

describe('OcTile component', () => {
  it('renders default space correctly', () => {
    const wrapper = getWrapper({ resource: getSpaceMock() })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders disabled space correctly', () => {
    const wrapper = getWrapper({ resource: getSpaceMock(true), isResourceDisabled: true })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders selected resource correctly', () => {
    const wrapper = getWrapper({ resource: getSpaceMock(), isResourceSelected: true })
    expect(wrapper.find('.oc-tile-card-selected').exists()).toBeTruthy()
  })
  it.each(['xlarge, xxlarge, xxxlarge'])(
    'renders resource icon size correctly',
    (resourceIconSize) => {
      const wrapper = getWrapper({ resource: getSpaceMock(), resourceIconSize })
      expect(wrapper.find('resource-icon-stub').attributes().size).toEqual(resourceIconSize)
    }
  )
  it('shows a loading spinner if isLoading is set to true', () => {
    const wrapper = getWrapper({ resource: getSpaceMock(), isLoading: true })
    expect(wrapper.find('.oc-tile-card-loading-spinner').exists()).toBeTruthy()
  })

  function getWrapper(props = {}) {
    return shallowMount(ResourceTile, {
      props,
      global: { plugins: [...defaultPlugins()], renderStubDefaultSlot: true }
    })
  }
})
