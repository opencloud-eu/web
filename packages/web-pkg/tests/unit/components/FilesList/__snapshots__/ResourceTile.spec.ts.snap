// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OcTile component > renders default space correctly 1`] = `
"<div class="oc-tile-card oc-card flex flex-col h-full shadow-none bg-role-surface-container hover:bg-role-surface-container-highest outline outline-role-surface-container-highest">
  <resource-link-stub resource="[object Object]" isresourceclickable="true" class="oc-card-media-top flex justify-center items-center m-0 w-full relative aspect-[16/9]" tabindex="-1">
    <div class="oc-tile-card-selection absolute top-0 left-0 [&amp;_input]:bg-role-surface-container"></div>
    <!--v-if-->
    <div class="oc-tile-card-preview flex items-center justify-center text-center size-full absolute">
      <resource-icon-stub resource="[object Object]" size="xlarge" class="tile-default-image pt-1 relative"></resource-icon-stub>
    </div>
  </resource-link-stub>
  <div class="oc-card-body p-2">
    <div class="flex justify-between items-center">
      <div class="flex items-center truncate resource-name-wrapper text-role-on-surface overflow-hidden">
        <resource-list-item-stub resource="[object Object]" pathprefix="" ispathdisplayed="false" parentfoldername="" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="true" isicondisplayed="false" isresourceclickable="true"></resource-list-item-stub>
      </div>
      <div class="flex items-center">
        <!-- Slot for indicators !-->
        <!-- Slot for individual actions -->
        <!-- Slot for contextmenu -->
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`OcTile component > renders disabled space correctly 1`] = `
"<div class="oc-tile-card oc-card flex flex-col h-full shadow-none bg-role-surface-container hover:bg-role-surface-container-highest outline outline-role-surface-container-highest oc-tile-card-disabled opacity-70 grayscale-60 pointer-events-none">
  <resource-link-stub resource="[object Object]" isresourceclickable="true" class="oc-card-media-top flex justify-center items-center m-0 w-full relative aspect-[16/9]" tabindex="-1">
    <div class="oc-tile-card-selection absolute top-0 left-0 [&amp;_input]:bg-role-surface-container"></div>
    <!--v-if-->
    <div class="oc-tile-card-preview flex items-center justify-center text-center size-full absolute">
      <resource-icon-stub resource="[object Object]" size="xlarge" class="tile-default-image pt-1 relative"></resource-icon-stub>
    </div>
  </resource-link-stub>
  <div class="oc-card-body p-2">
    <div class="flex justify-between items-center">
      <div class="flex items-center truncate resource-name-wrapper text-role-on-surface overflow-hidden">
        <resource-list-item-stub resource="[object Object]" pathprefix="" ispathdisplayed="false" parentfoldername="" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="true" isicondisplayed="false" isresourceclickable="true"></resource-list-item-stub>
      </div>
      <div class="flex items-center">
        <!-- Slot for indicators !-->
        <!-- Slot for individual actions -->
        <!-- Slot for contextmenu -->
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
